tools/dissector/sample_dissector.cpp

```diff
@@ -31,13 +32,171 @@ namespace OpenDDS
   namespace DCPS
   {
 
+    bool utf16_to_utf8(
+      std::string & to, ACE_CDR::WChar* from, size_t length = 1
+    ) {
+      size_t size = length * sizeof(ACE_CDR::WChar);
+      gchar* from_;
+
+#if ACE_SIZEOF_WCHAR == 4
+      /*
+       * We have to trim the extra 16 bits of zeros between the characters
+       * before we can pass it to glib's g_convert. Otherwise it will
+       * interpret them as NULL characters and stop after the first two bytes.
+       * g_convert doesn't respsect the size passed to it if it thinks it can
+       * keep going.
+       */
+      uint16_t* trimmed = new uint16_t[length + 1];
+      for (size_t i = 0; i < length; i++) {
+        trimmed[i] = from[i];
+      }
+      trimmed[length] = 0;
+      from_ = reinterpret_cast<gchar*>(trimmed);
+#else
+      from_ = reinterpret_cast<gchar*>(from);
+#endif
+
+      const char * error_msg = "UTF-16 to UTF-8 conversion failed: ";
+      GError * error = NULL;
+      gsize bytes_read = 0;
+      gsize bytes_written = 0;
+      char * utf8 = g_convert(
+        from_, size,
+        "UTF-8", "UTF-16", &bytes_read, &bytes_written, &error
+      );
+
+      if (utf8 != NULL) {
+        to = utf8;
+        g_free(utf8);
+      } else {
+        ACE_DEBUG((LM_DEBUG, ACE_TEXT("%s%s\n"),
+          error_msg,
+          error->message
+        ));
+        to = error_msg;
+        to += error->message;
+        g_clear_error(&error);
+      }
+
+#if ACE_SIZEOF_WCHAR == 4
+      delete [] trimmed;
+#endif
+
+      return utf8 != NULL;
+    }
+
+    Wireshark_Bundle::Wireshark_Bundle(
+      char * data, size_t size, bool swap_bytes, Serializer::Alignment align
+    ) :
+      block(data, size),
+      serializer(&block, swap_bytes, align)
+    {
+      block.wr_ptr(data + size);
+      get_size_only = false;
+    }
+
+    Wireshark_Bundle::Wireshark_Bundle(const Wireshark_Bundle & other) :
+      block(other.block.rd_ptr(), other.block.size()),
+      serializer(
+        &block,
+        other.serializer.swap_bytes(),
+        other.serializer.alignment()
+      )
+    {
+      block.wr_ptr(other.block.wr_ptr() - other.block.rd_ptr());
+
+      get_size_only = other.get_size_only;
+      tvb = other.tvb;
+      info = other.info;
+      offset = other.offset;
+      use_index = other.use_index;
+      index = other.index;
+    }
+
+    size_t Wireshark_Bundle::buffer_pos() {
+      return reinterpret_cast<size_t>(block.rd_ptr());
+    }
+
     guint8 *
-    Wireshark_Bundle::get_remainder ()
+    Wireshark_Bundle::get_remainder()
     {
-      gint remainder = tvb_length(tvb) - offset;
+      gint remainder = ws_tvb_length(tvb) - offset;
       return reinterpret_cast<guint8 *>(ws_ep_tvb_memdup(tvb, offset, remainder));
     }
 
+    Sample_Base::~Sample_Base() {
+      for (
+        Field_Contexts::iterator i = field_contexts_.begin();
+        i != field_contexts_.end();
+        ++i
+      ) {
+        delete i->second;
+        i->second = NULL;
```

Comment:
Is there a need for this line?

Reply:
No, I guess I was just following the rule to NULL unused pointers, but it is a moot point here.